File: year/2025/day05/task1/main.py
"""
Author: Jose A. Romero
Puzzle: Advent of Code (year=2025 ; day=5 ; task=1)
"""

import sys


def main() -> None:
    inp_ranges, ids = sys.stdin.read().split("\n\n")
    inp_ranges = sorted([*map(int, i.split("-"))] for i in inp_ranges.splitlines())
    ids = sorted([*map(int, ids.splitlines())])

    ranges = [inp_ranges[0]]
    for l, r in inp_ranges[1:]:
        if l <= ranges[-1][1]:
            ranges[-1][1] = max(ranges[-1][1], r)

        else:
            ranges.append([l, r])

    ans = 0
    i, p = 0, 0
    while i < len(ids) and p < len(ranges):
        if ids[i] < ranges[p][0]:
            i += 1

        elif ranges[p][0] <= ids[i] <= ranges[p][1]:
            i += 1
            ans += 1

        else:  # ranges[p][1] < ids[i]
            p += 1

    print(ans)


if __name__ == "__main__":
    main()

File: year/2025/day05/task1/main.cpp
/*
Author: Jose A. Romero
Puzzle: Advent of Code (year=2025 ; day=5 ; task=1)
*/

#include <bits/stdc++.h>

#define ll long long
#define endl '\n'

using namespace std;

int main() {
    ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0);

    vector<pair<ll, ll>> inp_ranges;
    vector<ll> ids;

    bool read_ranges = true;
    for (string line ; getline(cin, line) ; ) {
        if (line.empty()) {
            read_ranges = false;
            continue;
        }
        if (read_ranges) {
            ll l, r;
            sscanf(line.c_str(), "%lld-%lld", &l, &r);
            inp_ranges.push_back({l, r});
        }
        else {
            ids.push_back(stoll(line));
        }
    }

    sort(inp_ranges.begin(), inp_ranges.end());
    vector<pair<ll, ll>> ranges = {inp_ranges[0]};
    for (int i = 1 ; i < inp_ranges.size() ; ++i) {
        if (inp_ranges[i].first <= ranges.back().second) {
            ranges.back().second = max(ranges.back().second, inp_ranges[i].second);
        }
        else {
            ranges.push_back(inp_ranges[i]);
        }
    }

    sort(ids.begin(), ids.end());

    int ans = 0;
    for (int i = 0, p = 0 ; i < ids.size() && p < ranges.size() ; ) {
        if (ids[i] < ranges[p].first) {
            ++i;
        }
        else if (ranges[p].first <= ids[i] && ids[i] <= ranges[p].second) {
            ++i;
            ++ans;
        }
        else {  // ranges[p].second < ids[i]
            ++p;
        }
    }
    cout << ans << endl;
    return 0;
}
